File: clients/ios/PcmRingBuffer.swift
import Foundation

/// Thread-safe 1.5 second PCM pre-roll at 20 ms per frame.
final class PcmRingBuffer {
    private let lock = NSLock()
    private var frames: [Data] = []
    private let maxFrames: Int
    private var streaming = false

    init(maxFrames: Int = 75) {
        self.maxFrames = maxFrames
    }

    /// Returns true when the caller should also forward this frame live.
    @discardableResult
    func append(_ frame: Data) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if frames.count == maxFrames { frames.removeFirst() }
        frames.append(frame)
        return streaming
    }

    func beginStream() -> [Data] {
        lock.lock()
        defer { lock.unlock() }
        let buffered = frames
        frames.removeAll(keepingCapacity: true)
        streaming = true
        return buffered
    }

    func stopStream() {
        lock.lock()
        defer { lock.unlock() }
        streaming = false
        frames.removeAll(keepingCapacity: true)
    }
}

File: clients/android/app/src/main/java/com/example/iotvoice/PcmRingBuffer.kt
package com.example.iotvoice

import java.util.ArrayDeque

/** Thread-safe 1.5 second PCM pre-roll at 20 ms per frame. */
class PcmRingBuffer(
    private val maxFrames: Int = 75,
) {
    private val frames = ArrayDeque<ByteArray>(maxFrames)
    private val lock = Any()
    private var streaming = false

    /** Returns true when the caller should also forward this frame live. */
    fun append(frame: ByteArray): Boolean = synchronized(lock) {
        if (frames.size == maxFrames) frames.removeFirst()
        frames.addLast(frame.copyOf())
        streaming
    }

    fun beginStream(): List<ByteArray> = synchronized(lock) {
        val buffered = frames.toList()
        frames.clear()
        streaming = true
        buffered
    }

    fun stopStream() = synchronized(lock) {
        streaming = false
        frames.clear()
    }
}
